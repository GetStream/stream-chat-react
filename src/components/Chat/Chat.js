// @ts-check

import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';

import Dayjs from 'dayjs';
import { ChatContext, TranslationContext } from '../../context';
import { Streami18n } from '../../i18n';

import { version } from '../../../package.json';

/**
 * Chat - Wrapper component for Chat. The needs to be placed around any other chat components.
 * This Chat component provides the ChatContext to all other components.
 *
 * The ChatContext provides the following props:
 *
 * - client (the client connection)
 * - channels (the list of channels)
 * - setActiveChannel (a function to set the currently active channel)
 * - channel (the currently active channel)
 *
 * It also exposes the withChatContext HOC which you can use to consume the ChatContext
 *
 * @example ../../docs/Chat.md
 * @typedef {import('stream-chat').Channel | undefined} ChannelState
 * @type {React.FC<{ client: import('types').StreamChatReactClient, theme?: string, i18nInstance?: Streami18n, initialNavOpen?: boolean }>}
 */
const Chat = ({
  client,
  theme = 'messaging light',
  i18nInstance,
  initialNavOpen = true,
  children,
}) => {
  const [translators, setTranslators] = useState(
    /** @type { Required<import('types').TranslationContextValue>} */ ({
      t: /** @param {string} key */ (key) => key,
      tDateTimeParser: (input) => Dayjs(input),
    }),
  );
  const [mutes, setMutes] = useState(
    /** @type {import('stream-chat').Mute[]} */ ([]),
  );
  const [navOpen, setNavOpen] = useState(initialNavOpen);
  const [channel, setChannel] = useState(
    /** @type {ChannelState} */ (undefined),
  );

  const openMobileNav = () => setTimeout(() => setNavOpen(true), 100);
  const closeMobileNav = () => setNavOpen(false);
  const clientMutes = client?.user?.mutes;

  useEffect(() => {
    if (!client) return;
    const userAgent = client.getUserAgent();
    if (!userAgent.includes('stream-chat-react')) {
      // should result in something like:
      // 'stream-chat-react-2.3.2-stream-chat-javascript-client-browser-2.2.2'
      client.setUserAgent(`stream-chat-react-${version}-${userAgent}`);
    }
    // don't want client in dep array because it is a required
    // prop for this component and we only want this run on mount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [client]);

  useEffect(() => {
    setMutes(clientMutes || []);

    /** @param {import('stream-chat').Event} e */
    const handleEvent = (e) => {
      if (e.type === 'notification.mutes_updated') setMutes(e.me?.mutes || []);
    };
    if (client) client.on(handleEvent);
    return () => client && client.off(handleEvent);
  }, [client, clientMutes]);

  useEffect(() => {
    let streami18n;
    if (i18nInstance instanceof Streami18n) streami18n = i18nInstance;
    else streami18n = new Streami18n({ language: 'en' });

    streami18n.registerSetLanguageCallback((t) =>
      setTranslators((prevTranslator) => ({ ...prevTranslator, t })),
    );
    streami18n.getTranslators().then((translator) => {
      if (translator) setTranslators(translator);
    });
  }, [i18nInstance]);

  const setActiveChannel = useCallback(
    /**
     * @param {ChannelState} activeChannel
     * @param {{ limit?: number; offset?: number }} [watchers]
     * @param {React.BaseSyntheticEvent} [e]
     */
    async (activeChannel, watchers = {}, e) => {
      if (e && e.preventDefault) e.preventDefault();

      if (activeChannel && Object.keys(watchers).length) {
        await activeChannel.query({ watch: true, watchers });
      }
      setChannel(activeChannel);
      closeMobileNav();
    },
    [],
  );

  if (!translators.t) return null;

  return (
    <ChatContext.Provider
      value={{
        client,
        theme,
        channel,
        mutes,
        navOpen,
        setActiveChannel,
        openMobileNav,
        closeMobileNav,
      }}
    >
      <TranslationContext.Provider value={translators}>
        {children}
      </TranslationContext.Provider>
    </ChatContext.Provider>
  );
};

Chat.propTypes = {
  /** The StreamChat client object */
  client: /** @type {PropTypes.Validator<import('stream-chat').StreamChat>} */ (PropTypes
    .object.isRequired),
  /**
   *
   * Theme could be used for custom styling of the components.
   *
   * You can override the classes used in our components under parent theme class.
   *
   * e.g. If you want to build a theme where background of message is black
   *
   * ```
   *  <Chat client={client} theme={demo}>
   *    <Channel>
   *      <MessageList />
   *    </Channel>
   *  </Chat>
   * ```
   *
   * ```scss
   *  .demo.str-chat {
   *    .str-chat__message-simple {
   *      &-text-inner {
   *        background-color: black;
   *      }
   *    }
   *  }
   * ```
   *
   * Built in available themes:
   *
   *  - `messaging light`
   *  - `messaging dark`
   *  - `team light`
   *  - `team dark`
   *  - `commerce light`
   *  - `commerce dark`
   *  - `livestream light`
   *  - `livestream dark`
   */
  theme: PropTypes.string,
  /** navOpen initial status */
  initialNavOpen: PropTypes.bool,
};

export default Chat;
