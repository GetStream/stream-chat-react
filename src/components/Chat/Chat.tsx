import React, { PropsWithChildren } from 'react';

import { useChat } from './hooks/useChat';
import { useCreateChatContext } from './hooks/useCreateChatContext';
import { useChannelsQueryState } from './hooks/useChannelsQueryState';
import { CustomStyles, darkModeTheme, useCustomStyles } from './hooks/useCustomStyles';

import { ChatProvider, CustomClasses, ThemeVersion } from '../../context/ChatContext';
import { SupportedTranslations, TranslationProvider } from '../../context/TranslationContext';

import type { StreamChat } from 'stream-chat';

import type { Streami18n } from '../../i18n/Streami18n';

import type { DefaultStreamChatGenerics } from '../../types/types';

/**
 * @deprecated will be removed with the complete transition to the theming V2 (next major release - `v11.0.0`)
 */
export type Theme<T extends string = string> =
  | 'commerce dark'
  | 'commerce light'
  | 'livestream dark'
  | 'livestream light'
  | 'messaging dark'
  | 'messaging light'
  | 'team dark'
  | 'team light'
  | T;

export type ChatProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  /** The StreamChat client object */
  client: StreamChat<StreamChatGenerics>;
  /** Object containing custom CSS classnames to override the library's default container CSS */
  customClasses?: CustomClasses;
  /**
   * @desc object containing custom styles to override the default CSS variables
   * @deprecated will be removed with the complete transition to the theming v2 (next major release - `v11.0.0`)
   */
  customStyles?: CustomStyles;
  /**
   * @desc if true, toggles the CSS variables to the default dark mode color palette
   * @deprecated will be removed with the complete transition to the theming v2 (next major release - `v11.0.0`)
   */
  darkMode?: boolean;
  /** Sets the default fallback language for UI component translation, defaults to 'en' for English */
  defaultLanguage?: SupportedTranslations;
  /** Instance of Stream i18n */
  i18nInstance?: Streami18n;
  /** Initial status of mobile navigation */
  initialNavOpen?: boolean;
  /** Used for injecting className/s to the Channel and ChannelList components */
  theme?: string;
  /** Windows 10 does not support country flag emojis out of the box. It chooses to render these emojis as characters instead. Stream
   * Chat can override this behavior by loading a custom web font that will render images instead (PNGs or SVGs depending on the platform).
   * Set this prop to true if you want to use these custom emojis for Windows users.
   */
  useImageFlagEmojisOnWindows?: boolean;
};

/**
 * Wrapper component for a StreamChat application. Chat needs to be placed around any other chat components
 * as it provides the ChatContext.
 */
export const Chat = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  props: PropsWithChildren<ChatProps<StreamChatGenerics>>,
) => {
  const {
    children,
    client,
    customClasses,
    customStyles,
    darkMode = false,
    defaultLanguage,
    i18nInstance,
    initialNavOpen = true,
    theme = 'messaging light',
    useImageFlagEmojisOnWindows = false,
  } = props;

  const {
    channel,
    closeMobileNav,
    getAppSettings,
    latestMessageDatesByChannels,
    mutes,
    navOpen,
    openMobileNav,
    setActiveChannel,
    translators,
  } = useChat({ client, defaultLanguage, i18nInstance, initialNavOpen });

  const channelsQueryState = useChannelsQueryState();
  const themeVersion: ThemeVersion =
    typeof window !== 'undefined'
      ? ((window
          .getComputedStyle(document.documentElement)
          .getPropertyValue('--str-chat__theme-version')
          .replace(' ', '') || '1') as ThemeVersion)
      : '1';

  useCustomStyles(darkMode ? darkModeTheme : customStyles);

  const chatContextValue = useCreateChatContext({
    channel,
    channelsQueryState,
    client,
    closeMobileNav,
    customClasses,
    getAppSettings,
    latestMessageDatesByChannels,
    mutes,
    navOpen,
    openMobileNav,
    setActiveChannel,
    theme,
    themeVersion,
    useImageFlagEmojisOnWindows,
  });

  if (!translators.t) return null;

  return (
    <ChatProvider value={chatContextValue}>
      <TranslationProvider value={translators}>{children}</TranslationProvider>
    </ChatProvider>
  );
};
