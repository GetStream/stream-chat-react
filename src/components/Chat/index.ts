export * from './Chat';
export * from './hooks/useChat';
export * from './hooks/useCustomStyles';
export * from './hooks/useCreateChatClient';
