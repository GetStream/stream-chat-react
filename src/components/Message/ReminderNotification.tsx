import React from 'react';
import { useChatContext, useTranslationContext } from '../../context';
import { useStateStore } from '../../store';
import type { Reminder, ReminderState } from 'stream-chat';

export type ReminderNotificationProps = {
  reminder?: Reminder;
};

const reminderStateSelector = (state: ReminderState) => ({
  timeLeftMs: state.timeLeftMs,
});

export const ReminderNotification = ({ reminder }: ReminderNotificationProps) => {
  const { client } = useChatContext();
  const { t } = useTranslationContext();
  const { timeLeftMs } = useStateStore(reminder?.state, reminderStateSelector) ?? {};

  const stopRefreshBoundaryMs = client.reminders.stopTimerRefreshBoundaryMs;
  const stopRefreshTimeStamp =
    reminder?.remindAt && stopRefreshBoundaryMs
      ? reminder?.remindAt.getTime() + stopRefreshBoundaryMs
      : undefined;

  const isBehindRefreshBoundary =
    !!stopRefreshTimeStamp && new Date().getTime() > stopRefreshTimeStamp;

  return (
    <p className='str-chat__message-reminder'>
      <span>{t<string>('Saved for later')}</span>
      {reminder?.remindAt && timeLeftMs !== null && (
        <>
          <span> | </span>
          <span>
            {isBehindRefreshBoundary
              ? t<string>('Due since {{ dueSince }}', {
                  dueSince: t<string>(`timestamp/ReminderNotification`, {
                    timestamp: reminder.remindAt,
                  }),
                })
              : t<string>(`Due {{ timeLeft }}`, {
                  timeLeft: t<string>('duration/Message reminder', {
                    milliseconds: timeLeftMs,
                  }),
                })}
          </span>
        </>
      )}
    </p>
  );
};
