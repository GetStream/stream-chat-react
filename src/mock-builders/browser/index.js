export * from './AnalyserNode';
export * from './AudioContext';
export * from './EventEmitter';
export * from './MediaRecorder';
export * from './ResizeObserver';
