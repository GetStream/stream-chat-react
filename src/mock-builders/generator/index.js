export * from './attachment';
export * from './channel';
export * from './emoji';
export * from './member';
export * from './message';
export * from './reaction';
export * from './user';
export * from './upload';
