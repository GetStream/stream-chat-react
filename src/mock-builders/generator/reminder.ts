import type { ReminderResponse } from 'stream-chat';

const baseData = {
  channel_cid: 'channel_cid',
  message_id: 'message_id',
  user_id: 'user_id',
} as const;

export const generateReminderResponse = ({
  data,
  scheduleOffsetMs,
}: {
  data?: Partial<ReminderResponse>;
  scheduleOffsetMs?: number;
} = {}): ReminderResponse => {
  const created_at = new Date().toISOString();
  const basePayload: ReminderResponse = {
    ...baseData,
    created_at,
    message: { id: baseData.message_id, type: 'regular' },
    updated_at: created_at,
    user: { id: baseData.user_id },
  };
  if (typeof scheduleOffsetMs === 'number') {
    basePayload.remind_at = new Date(
      new Date(created_at).getTime() + scheduleOffsetMs,
    ).toISOString();
  }
  return {
    ...basePayload,
    ...data,
  };
};
