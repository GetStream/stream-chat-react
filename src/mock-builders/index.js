/* eslint-disable no-param-reassign */
/* eslint-disable no-underscore-dangle */
import { StreamChat } from 'stream-chat';
import { nanoid } from 'nanoid';

const apiKey = 'API_KEY';
const token = 'dummy_token';

const connectUser = (client, user) =>
  new Promise((resolve) => {
    client.connectionId = 'dumm_connection_id';
    client.user = user;
    client.user.mutes = [];
    client._user = { ...user };
    client.userID = user.id;
    client.userToken = token;
    client.wsPromise = Promise.resolve(true);
    resolve();
  });

function mockClient(client) {
  jest.spyOn(client, '_setToken').mockImplementation();
  jest.spyOn(client, '_setupConnection').mockImplementation();
  jest.spyOn(client, '_setupConnection').mockImplementation();
  jest.spyOn(client, 'getAppSettings').mockImplementation();
  client.tokenManager = {
    getToken: jest.fn(() => token),
    tokenReady: jest.fn(() => true),
  };
  client.connectUser = connectUser.bind(null, client);
  return client;
}

export const getTestClient = () => mockClient(new StreamChat(apiKey));

export const getTestClientWithUser = async (user = { id: nanoid() }) => {
  const client = mockClient(new StreamChat(apiKey));
  await connectUser(client, user);
  return client;
};

export const getRandomInt = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min; // The maximum is exclusive and the minimum is inclusive
};

export * from './api';
export * from './event';
export * from './generator';
export * from './translator';
export * from './utils';
