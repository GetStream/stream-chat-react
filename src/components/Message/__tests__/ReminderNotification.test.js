import React from 'react';
import { act, render } from '@testing-library/react';
import { Chat } from '../../Chat';
import { ReminderNotification } from '../ReminderNotification';
import { generateUser, getTestClientWithUser } from '../../../mock-builders';
import { generateReminderResponse } from '../../../mock-builders/generator/reminder';

const user = generateUser();
const renderComponent = async ({ client, reminder }) => {
  let result;
  await act(() => {
    result = render(
      <Chat client={client}>
        <ReminderNotification reminder={reminder} />
      </Chat>,
    );
  });
  return result;
};

describe('ReminderNotification', () => {
  it('displays text for bookmark notifications', async () => {
    const client = await getTestClientWithUser(user);
    const reminderResponse = generateReminderResponse();
    client.reminders.upsertToState({
      data: reminderResponse,
    });
    const reminder = client.reminders.getFromState(reminderResponse.message_id);
    const { container } = await renderComponent({ client, reminder });
    expect(container).toMatchSnapshot();
  });
  it('displays text for bookmark notifications and time due in case of timed reminders', async () => {
    const client = await getTestClientWithUser(user);
    const reminderResponse = generateReminderResponse({
      scheduleOffsetMs: 60 * 1000,
    });
    client.reminders.upsertToState({
      data: reminderResponse,
    });
    const reminder = client.reminders.getFromState(reminderResponse.message_id);
    const { container } = await renderComponent({ client, reminder });
    expect(container).toMatchSnapshot();
  });
  it('displays text for bookmark notifications and reminder deadline if trespassed the refresh boundary', async () => {
    const client = await getTestClientWithUser(user);
    const reminderResponse = generateReminderResponse({
      data: { remind_at: new Date(0).toISOString() },
    });
    client.reminders.upsertToState({
      data: reminderResponse,
    });
    const reminder = client.reminders.getFromState(reminderResponse.message_id);
    const { container } = await renderComponent({ client, reminder });
    expect(container).toMatchSnapshot();
  });
});
