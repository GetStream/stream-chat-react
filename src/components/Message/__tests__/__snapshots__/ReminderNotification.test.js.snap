// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReminderNotification displays text for bookmark notifications 1`] = `
<div>
  <p
    class="str-chat__message-reminder"
  >
    <span>
      Saved for later
    </span>
  </p>
</div>
`;

exports[`ReminderNotification displays text for reminder deadline if trespassed the refresh boundary 1`] = `
<div>
  <p
    class="str-chat__message-reminder"
  >
    <span>
      Saved for later
    </span>
    <span>
       | 
    </span>
    <span>
      Due since 01/01/1970
    </span>
  </p>
</div>
`;

exports[`ReminderNotification displays text for time due in case of timed reminders 1`] = `
<div>
  <p
    class="str-chat__message-reminder"
  >
    <span>
      Saved for later
    </span>
    <span>
       | 
    </span>
    <span>
      Due in a minute
    </span>
  </p>
</div>
`;
